src/PluginHost.cpp

```diff
@@ -0,0 +1,585 @@
+   {
+      mCallback = callback;
+      mState = State::Connect;
+
+      Start(1);//Timer
+   }
+
+   void Finalize() override
+   {
+      Stop();//Timer
+
+      mCallback = nullptr;
+      
+      mServer.reset();
+   }
+   
+private:
+   
+   void StartHost()
+   {
+      auto process = std::make_unique<wxProcess>();
+      process->Detach();
+      if(LaunchHostProcess(process.get()))
+         //process will delete itself upon termination
+         process.release();
+      else
+      {
+         Stop();//Timer
+         mCallback->OnValidatorError("Cannot start plugin host");
+      }
+   }
+};
+
+
+class PluginHost : public IPCChannelStatusCallback
+{
+   std::unique_ptr<IPCClient> mClient;
+   IPCChannel* mChannel{nullptr};
+   InputMessageReader mInputMessageReader;
+   spinlock mSync;
+   std::unique_ptr<wxString> mRequest;
+   bool mRunning{true};
+
+public:
+
+   PluginHost()
+   {
+      AudacityApp::InitPluginHost();
+      mClient = std::make_unique<IPCClient>();
+      mClient->CreateChannel(4242, this);
+   }
+   
+   void OnConnect(IPCChannel* channel) override
+   {
+      std::lock_guard lck(mSync);
+      mChannel = channel;
+   }
+
+   void OnConnectionError() override
+   {
+      std::lock_guard lck(mSync);
+      mRunning = false;
+   }
+
+   void OnDataAvailable(const void* bytes, size_t length) override
+   {
+      mInputMessageReader.ConsumeBytes(bytes, length);
+      if(mInputMessageReader.CanPop())
+      {
+         auto request = std::make_unique<wxString>(mInputMessageReader.Pop());
+
+         std::lock_guard lck(mSync);
+         assert(!mRequest);
+         mRequest = std::move(request);
+      }
+   }
+
+   void OnDisconnect() override
+   {
+      std::lock_guard lck(mSync);
+      mChannel = nullptr;
+      mRunning = false;
+   }
+
+   bool Serve()
+   {
+      std::unique_ptr<wxString> request;
+      {
+         std::lock_guard lck(mSync);
+         if(!mRunning)
+            return false;
+         request = std::move(mRequest);
+      }
+      if(request)
+      {
+         wxString providerId;
+         wxString pluginPath;
+         wxString result;
+         if(ParseRequestString(*request, providerId, pluginPath))
+            result = Discover(providerId, pluginPath);
+         else
+            result = PluginsXMLHandler::MakeError("malformed request string");
+         std::lock_guard lck(mSync);
+         if(mChannel)
+            PutMessage(*mChannel, result);
+      }
+      return true;
+   }
+};
+
+class PluginHostModule final :
+   public wxModule
+{
+   static const wxString HostArgument;
+  
+public:
+   DECLARE_DYNAMIC_CLASS(PluginHostModule)
+
+   bool OnInit() override
+   {
+      if(wxTheApp && wxTheApp->argc >= 2 && wxStrcmp(wxTheApp->argv[1], HostArgument) == 0)
+      {
+         //log messages will appear in a separate window
+         //redirect to log file later
+         wxLog::EnableLogging(false);
+
+         PluginHost host;
+         while(host.Serve()) { }
```

Comment:
I am beginning to understand this new file.  If you are here, you are in the main thread of a child process.  The Serve() loop detects requests and sends responses, while the overridden functions do work on another thread.

Nothing ever blocks inside Serve(), which makes it a busy loop, in the admittedly unlikely event that it handles its requests faster than the responses arrive.  Whereas `OnDataAvailable` contains an assertion that is not provable from what I read here, and migh fail with unusual timing, unless there is more I don't understand yet about the main process.

I would be more satisfied to see:

- a std::mutex and associated std::condition_variable.
- Serve() waits https://en.cppreference.com/w/cpp/thread/condition_variable/wait for the condition: `!running || (request && channel)`
- OnConnectionError, OnDisconnect, an OnDataAvailable may make the condition true, so they call https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one
- Serve() releases the mutex only after moving from mRequest.
- OnDataAvailable holds the mutex while assigning to mRequest
  - After taking the mutex, it also waits on the condition variable for `!mRequest`, which really "proves" it
  - OnDisconnect, which can make the condition true, also notifies the variable